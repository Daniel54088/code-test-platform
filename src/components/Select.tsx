import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import CircularProgress from '@mui/material/CircularProgress';
import { CaptionProps, ListProps, SelectOption } from '../interfaces';
import { SelectProps } from '../interfaces';


const Select: React.FC<SelectProps> = (props) => {
    
    const [data, setData] = useState<Object[]>(props.data);

    const [isfocus, setIsfocus] = useState(false);
    const [selectedItem, setSelectedItem] = useState<SelectOption | null>(null);

    const handleSelectClick = () => {
        setIsfocus(true);
    }


    useEffect(() => {
        setData(props.data)
    }, [props.data]);

    const customFilter = (option: { label: String }, searchText: string) => {
        if (
            option.label.toLowerCase().includes(searchText.toLowerCase())) {
            return true;
        } else {
            return false;
        }
    }

    const filterDataList = (value: string) => {

        let filteredData = props.data.filter((item: any) => {
            return customFilter(item, value);
        })

        setData(filteredData);
    }

    const handleOptionClick = (option: any) => {
        setSelectedItem(option);
        setIsfocus(false);
        setData(props.data);
    }


    return (
        <OuterContainer >
            <span className="fi fi-gr"></span>
            <SelectContainer
                onClick={handleSelectClick}
                onFocus={() => setIsfocus(true)}
                onBlur={() => {
                    setIsfocus(false);
                    setData(props.data);
                }}
            >
                <Dropdown>
                    <Caption isFocus={isfocus}>
                        {isfocus ? (
                            <Input
                                onChange={(e) => filterDataList(e.target.value)}
                                autoFocus
                            />
                        ) : (
                            <React.Fragment>
                                {selectedItem === null ? (
                                    <UnFocusText>
                                        Select
                                    </UnFocusText>
                                ) :
                                    (
                                        <SelectedOption>
                                            <span style={{ paddingRight: '1rem' }}>{selectedItem.pngUrl}</span>
                                            <ItemText>{selectedItem.label}</ItemText>
                                        </SelectedOption>
                                    )}
                            </React.Fragment>
                        )}

                        <KeyboardArrowDownIcon style={{ fill: '#697689' }} />
                    </Caption>

                </Dropdown>

            </SelectContainer>

            <List isFocus={isfocus}>
                {
                    props.loading
                        ? (
                            <SpinnerBox>
                                <CircularProgress />
                            </SpinnerBox>
                        )
                        : data.map((item: any, idx) =>
                            <Item key={idx}
                                onMouseDown={() => handleOptionClick(item)}>
                                <span style={{ paddingRight: '1rem' }}>{item.pngUrl}</span>
                                <ItemText>{item.label}</ItemText>
                            </Item>
                        )}

            </List>
        </OuterContainer>
    );
}

export default Select;

const OuterContainer = styled.div`
    display: flex; 
    flex-direction: column;  
`;

const SelectContainer = styled.div.attrs({ 'aria-label': 'select-container' })`
    width: 260px;
    background-color: transparent;
    magin-top: 8rem;
`;

const Dropdown = styled.div`

`;

const Caption = styled.div<CaptionProps>`
    border-color: ${props => props.isFocus ? "#6200EE" : "rgb(204, 204, 204)"};
    border-width: ${props => props.isFocus ? "2px" : "1px;"};
    border-style: solid;   
    padding: 11px 12px;
    border-radius: 6px;
    cursor: pointer;
    display: flex;
`;

const UnFocusText = styled.div`
    width: 100%;
    font-weight: 500;
    font-size: 15px;
    line-height: 22px;
    color: #697689;
}
`;

const Input = styled.input.attrs({ placeholder: "Search",'aria-label': 'country-search' })`
    width: 100%;
    border:0;
    font-weight: 500;
    font-size: 15px;
    line-height: 22px;
    &:focus {
        outline: none;
    }
`;

const List = styled.div.attrs<ListProps>(
    ({ isFocus }) => ({
        'aria-label': 'List'
    })
  )<ListProps>`
    width: 260px;
    border-radius: 5px;
    display:${props => props.isFocus ? "block" : "none"};
    box-shadow: 0 1px 24px 4px rgb(0 0 0 / 13%);
    padding-top: 10px;
    padding-bottom: 10px;
    max-height: 242px;
    overflow:auto;
`;

const Item = styled.div.attrs<any>(props => ({
    value: props.value
}))`
    padding: 11px 24px;
    cursor: pointer;
    width: 215px;
    display: flex;
    &:hover {
        background-color : #f5f5f5;
        color: #6200EE;
    }
`;

const ItemText = styled.span.attrs({ 'aria-label': 'country-name' })`
    text-overflow:ellipsis;
    overflow: hidden;
    white-space: nowrap;
    line-height: 23px
`;

const SelectedOption = styled.div`
    width: 100%;
    font-weight: 500;
    font-size: 15px;
    line-height: 22px;  
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
`;


const SpinnerBox = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    height: 200px;
`;