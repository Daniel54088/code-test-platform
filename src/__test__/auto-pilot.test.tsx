import React from 'react';
import {render, fireEvent,  screen, RenderResult} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import Select from '../components/Select';

const correctDataFomat =[
  {
    label: "Papua New Guinea",
    pngUrl: "ðŸ‡µðŸ‡¬",
    value: "Papua New Guinea"
  },
  {
    label: "Lithuania",
    pngUrl: "ðŸ‡±ðŸ‡¹",
    value: "Lithuania",
  }
];

let documentBody: RenderResult;

describe('<Select />', () => {

  beforeEach(() => {
    documentBody = render(<Select data={correctDataFomat} loading={false} />);
  });

  it('matches snapshot', () => {
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it('Shows Select when in the begin', () => {
    const SelectContainer = screen.getByLabelText('select-container');
    expect(SelectContainer).toHaveTextContent('Select')
  });

  it('List will display none before click and display block after click', () => {
   
    const SelectContainer = screen.getByLabelText('select-container');
    const List = screen.getByLabelText('List');
    expect(List).toHaveStyle(`display: none`);
    userEvent.click(SelectContainer);
    expect(List).toHaveStyle(`display: block`);

  });

  it('Country items will render successfully', () => {
   
    const SelectContainer = screen.getByLabelText('select-container');
    userEvent.click(SelectContainer);

    const countryItems = screen.getAllByLabelText('country-name');

    expect(countryItems[0]).toHaveTextContent('Papua New Guinea');
    expect(countryItems[1]).toHaveTextContent('Lithuania');

  });

  it('Search item works fine', () => {
    //Click Select
    const SelectContainer = screen.getByLabelText('select-container');
    userEvent.click(SelectContainer);
    //Insert text in input
    const searchInput = screen.getByLabelText('country-search');
    userEvent.type(searchInput, 'Pa');
    // Only one country should be display
    const countryItems = screen.getAllByLabelText('country-name');
    expect(countryItems.length).toBe(1);
    expect(countryItems[0]).toHaveTextContent('Papua New Guinea');
  });

});