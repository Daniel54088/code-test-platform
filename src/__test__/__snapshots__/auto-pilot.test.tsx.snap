// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="sc-bdvvtL iHzIin"
    >
      <span
        class="fi fi-gr"
      />
      <div
        aria-label="select-container"
        class="sc-gsDKAQ fVdBuD"
      >
        <div
          class="sc-dkPtRN iIDBAs"
        >
          <div
            class="sc-hKwDye GGwNJ"
          >
            <div
              class="sc-eCImPb htOSmG"
            >
              Select
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              style="fill: #697689;"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        aria-label="List"
        class="sc-gKclnd gSctZj"
      >
        <div
          class="sc-iCfMLu ehtORy"
        >
          <span
            style="padding-right: 1rem;"
          >
            ðŸ‡µðŸ‡¬
          </span>
          <span
            aria-label="country-name"
            class="sc-furwcr lbChbg"
          >
            Papua New Guinea
          </span>
        </div>
        <div
          class="sc-iCfMLu ehtORy"
        >
          <span
            style="padding-right: 1rem;"
          >
            ðŸ‡±ðŸ‡¹
          </span>
          <span
            aria-label="country-name"
            class="sc-furwcr lbChbg"
          >
            Lithuania
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
